### ISSUE TITLE
Question regarding the preprocessing of image

### ISSUE THREAD

**Issue by kennyvoo** [lucazanella/AnomalyCLIP#8]
Good work! May I know what's the preprocessing step to get feature that you've provided?
The following are the snippet of the code that I've used to extract the features, but I realized that the features extracted by the clip is not the same as what you've uploaded.

I've compared the features encoded by the clip and your uploaded feature but it's not the same.
[provided_feat.txt](https://github.com/luca-zanella-dvl/AnomalyCLIP/files/13903385/provided_feat.txt)
[clip_extracted.txt](https://github.com/luca-zanella-dvl/AnomalyCLIP/files/13903384/clip_extracted.txt)

```python
from src.models.components.anomaly_clip import AnomalyCLIP
import yaml
import torch
from PIL import Image
from src.utils.augmentations import get_augmentations
import glob
import math
import numpy as np
import imageio
from PIL import Image, ImageDraw

def round_to_nearest(number: float, X: int) -> int:
    """Rounds a number to the smallest upper integer that is divisible by a given integer X.

    Args:
    number: A float or integer to be rounded to the nearest multiple of X.
    X: An integer that the rounded number should be divisible by.

    Returns:
    The smallest upper integer that is divisible by X and is closest to the input number.
    """
    return math.ceil(number / X) * X

def load_yaml(path):
    with open(path, 'r') as yaml_file:
        config = yaml.full_load(yaml_file)
    config["net"]["load_from_features"]= False
    config["net"]["stride"] = 1
    config["net"]["ncrops"] = 1
    config["net"]["labels_file"] = "data/sht_labels.csv"
    config["net"]["normal_id"] = 8
    return config["net"]

def load_model(config):
    model = AnomalyCLIP(**config)
    state_dict = torch.load("checkpoints/shanghaitech/last.ckpt")["state_dict"]
    state_dict = {key.replace('net.', ''): value for key, value in state_dict.items()}

    model.load_state_dict(state_dict,strict=False)
    model.cuda()
    for p in model.image_encoder.parameters():
        p.requires_grad = False
    for p in model.text_encoder.parameters():
        p.requires_grad = False
    model.text_encoder.text_projection.requires_grad = True
    for p in model.token_embedding.parameters():
        p.requires_grad = False
    return model

if __name__ == "__main__":
    config = load_yaml("configs/model/anomaly_clip_shanghaitech.yaml")
    model = load_model(config)
    ncentroid = torch.load("logs/train/runs/shanghaitech/ncentroid.pt")

    images = []
    files_list = sorted(glob.glob("../DMAD/DMAD-PDM/data/shanghai/training/frames/01_001/*"), key=lambda x: int(x.split('/')[-1].split(".")[0]))
    num_segments = 32
    frames_per_segment = 16
    stride = 1
    end_frame = round_to_nearest(
        len(files_list),
        num_segments * frames_per_segment * stride,
    )
    start_indices = np.arange(end_frame / (frames_per_segment * stride)) * (
        frames_per_segment * stride
    )        
    for start_index in start_indices:
        for i in range(frames_per_segment):
            frame_index = (int(start_index) + i * stride) % len(files_list)
            frame_index += 0  # make sure to start from the correct frame
            images.append(Image.open(files_list[frame_index]).convert("RGB"))

    test_transform = get_augmentations(224, 1)
    images_feat = test_transform(images)
    images_feat = images_feat.cuda().unsqueeze(0)
    labels = torch.tensor([8]*512).cuda()

    # Forward pass
    similarity, abnormal_scores = model(
        images_feat,
        labels,
        ncentroid,
        1,
        test_mode=True,
    )
```

**Comment by luca-zanella-dvl** [comment]
Hi Kenny, thank you!

Thank you for sharing the code snippet. After adding the following lines of code:

```
data["net"]["stride"] = 1
data["net"]["ncrops"] = 1
data["net"]["labels_file"] = "data/sht_labels.csv"
data["net"]["normal_id"] = 8
```

I tested the modified snippet and I observed that the features extracted from [here](https://github.com/luca-zanella-dvl/AnomalyCLIP/blob/ffb0aa623e2db755c195f708eeba7fb49a1c4b01/src/models/components/anomaly_clip.py#L121) closely align with the ones provided in the Google Drive folder. The only notable difference is in the shape of the first dimension because of the padding applied in test mode (the shape is not [764, 512], where 764 is the number of frames and 512 is the embedding size, but [1024, 512]). 

Could you provide more details on how you computed `clip_extracted.txt`?

**Comment by kennyvoo** [comment]
I've kept the code the same except initializing clip to cuda.
L132 of  `anomaly_clip.py`, I print image_features[0][0][0]

**Comment by luca-zanella-dvl** [comment]
If you are only interested in the CLIP features, then you should use [L121](https://github.com/luca-zanella-dvl/AnomalyCLIP/blob/ffb0aa623e2db755c195f708eeba7fb49a1c4b01/src/models/components/anomaly_clip.py#L121C17-L123C69). The code at line 132 contains image features rearranged to accommodate videos with arbitrary lengths, enabling compatibility with our Temporal model.

**Comment by kennyvoo** [comment]
the result on L121 and L132 are the same since I'm only comparing the feature of the first frame

**Comment by luca-zanella-dvl** [comment]
A possibility is that the CLIP model is not loaded correctly. I am using the ViT-B/16 model, which should be automatically downloaded and stored at .cache/clip/ViT-B-16.pt. Could you try removing the cached version and then run the code again? This should download the model again.

**Comment by kennyvoo** [comment]
I've tried. still the same.
could it be the way I extracted the frames?
I used the following bash script to extract it.
```
for f in ./videos/*
do
   s=${f##*/}
   s=${s%.avi}
   echo $s
   mkdir ./frames/$s

   ffmpeg -i $f ./frames/$s/%04d.jpg
done

```

**Comment by luca-zanella-dvl** [comment]
Good point. I used OpenCV for frame extraction. As JPEG is a lossy format, the default quality setting in ffmpeg might differ from OpenCV's default quality. A solution similar to this should work.

```
import argparse
import os
from pathlib import Path
import cv2

def extract_frames(video_path, frames_dir):
    video_name = Path(video_path).stem

    video_frames_dir = os.path.join(frames_dir, video_name)
    os.makedirs(video_frames_dir, exist_ok=True)

    cap = cv2.VideoCapture(video_path)

    frame_count = 0
    while cap.isOpened():
        ret, frame = cap.read()
        if not ret:
            break

        frame_path = os.path.join(video_frames_dir, f"{frame_count:06d}.jpg")
        
        cv2.imwrite(frame_path, frame)

        frame_count += 1

    cap.release()
    print(f"Extracted {frame_count} frames from {video_path} to {video_frames_dir}")


def main(videos_dir, frames_dir):
    os.makedirs(frames_dir, exist_ok=True)

    # Loop through all the video files in the video directory
    for video_file in os.listdir(videos_dir):
        if video_file.endswith(".avi") or video_file.endswith(".mp4"):
            video_path = os.path.join(videos_dir, video_file)
            extract_frames(video_path, frames_dir)

def parse_args():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--videos_dir",
        type=str,
        required=True,
        help="Directory path to the videos.",
    )
    parser.add_argument(
        "--frames_dir",
        type=str,
        required=True,
        help="Directory path to the frames.",
    )
    args = parser.parse_args()
    return args


if __name__ == "__main__":
    args = parse_args()
    main(args.videos_dir, args.frames_dir)
```

**Comment by kennyvoo** [comment]
Thank you for supporting !!! It's indeed this issue.

**Comment by luca-zanella-dvl** [comment]
I am glad to hear that! Thanks for pointing this out, we will be sure to include this information in the readme.
